### PR TITLE
avatar-avatar collisions work again

### PR DESCRIPTION
**Pull request by AndrewMeadows**
Avatar's get an AvatarMotionState as their bridge object.  AvatarManager maintains the list of AvatarMotionStates.

Known issue: It isn't easy to **push** another avatar around, but your avatar will collide with theirs and stop.

### CODE REVIEW

File: libraries/avatars/src/AvatarHashMap.cpp [hunk 1]
```diff
@@ -20,19 +20,6 @@ AvatarHashMap::AvatarHashMap() {
...
-    qCDebug(avatars) << "Removing Avatar with UUID" << iterator.key() << "from AvatarHashMap.";
-    return _avatarHash.erase(iterator);
-}
-
-const qint64 AVATAR_SILENCE_THRESHOLD_USECS = 5 * USECS_PER_SECOND;
-
-bool AvatarHashMap::shouldKillAvatar(const AvatarSharedPointer& sharedAvatar) {
-    return (sharedAvatar->getOwningAvatarMixer() == NULL
-            || sharedAvatar->getUsecsSinceLastUpdate() > AVATAR_SILENCE_THRESHOLD_USECS);
-}
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarHashMap.cpp]: The intelligence here checking values of AvatarMixer and UsecsSinceLastUpdate moved into Avatar::shouldDie().

File: libraries/avatars/src/AvatarHashMap.cpp [hunk 2]
```diff
@@ -20,19 +20,6 @@ AvatarHashMap::AvatarHashMap() {
     connect(DependencyManager::get<NodeList>().data(), &NodeList::uuidChanged, this, &AvatarHashMap::sessionUUIDChanged);
 }
 
-
-AvatarHash::iterator AvatarHashMap::erase(const AvatarHash::iterator& iterator) {
-    qCDebug(avatars) << "Removing Avatar with UUID" << iterator.key() << "from AvatarHashMap.";
-    return _avatarHash.erase(iterator);
-}
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarHashMap.cpp]: AvatarHashMap::erase() moved to AvatarHashMap::remove().

File: libraries/physics/src/PhysicalEntitySimulation.h [hunk 3]
```diff
@@ -33,7 +30,7 @@ class PhysicalEntitySimulation :public EntitySimulation {
     PhysicalEntitySimulation();
     ~PhysicalEntitySimulation();
 
-    void init(EntityTree* tree, PhysicsEngine* engine, ShapeManager* shapeManager, EntityEditPacketSender* packetSender);
+    void init(EntityTree* tree, PhysicsEngine* engine, EntityEditPacketSender* packetSender);
```
Comment (AndrewMeadows) [inline libraries/physics/src/PhysicalEntitySimulation.h]: PhysicalEntitySimulation no longer needs to know about the ShapeManager.  Instead we use the static accessor to it in ObjectMotionState, the base class for all of our MotionStates, and all shape changes happen inside the MotionState context.

File: libraries/physics/src/ShapeFactory.cpp [hunk 4]
```diff
@@ -95,5 +98,17 @@ btCollisionShape* ShapeInfoUtil::createShapeFromInfo(const ShapeInfo& info) {
...
+            auto compound = new btCompoundShape(); 
+            btTransform trans;
+            trans.setIdentity();
+            trans.setOrigin(glmToBullet(info.getOffset()));
+            compound->addChildShape(trans, shape);
+            shape = compound;
+        }
+    }
     return shape;
 }
```
Comment (AndrewMeadows) [inline libraries/physics/src/ShapeFactory.cpp]: In theory this code will support entities with non-trivial "registration point" however I think I still need to connect a few pipes to make it work.  The effect of wrapping a regular shape inside a btCompoundShape is that the center of mass of the final object will remain at the "object's" position (e.g. the Entity's  or Avatar's position).

File: libraries/shared/src/ShapeInfo.cpp [hunk 5]
```diff
@@ -201,8 +213,12 @@ const DoubleHashKey& ShapeInfo::getHash() const {
...
-                DoubleHashKey::hashFunction2((uint32_t)(_halfExtents[j] * MILLIMETERS_PER_METER + copysignf(1.0f, _halfExtents[j]) * 0.49f));
+            uint32_t floatHash = DoubleHashKey::hashFunction2(
+                    (uint32_t)(_halfExtents[j] * MILLIMETERS_PER_METER + copysignf(1.0f, _halfExtents[j]) * 0.49f));
+            if (useOffset) {
+                floatHash ^= DoubleHashKey::hashFunction2(
+                        (uint32_t)(_offset[j] * MILLIMETERS_PER_METER + copysignf(1.0f, _offset[j]) * 0.49f));
+            }
             hash += ~(floatHash << 17);
             hash ^=  (floatHash >> 11);
             hash +=  (floatHash << 4);
```
Comment (AndrewMeadows) [inline libraries/shared/src/ShapeInfo.cpp]: Adding support for offset inside the ShapeInfo hash key generation, which means the ShapeManager should be able to share shapes with offset.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 461d08c0a953a2cbd7200775e4bbf4f652247240 available at: https://gist.github.com/56b44e3f86e1715cfed3<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4944.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4944.dmg

**Pull request merged**